### PR TITLE
  EIP-6780: SELFDESTRUCT only in same transaction

### LINKED ISSUE #524: EIP-6780: SELFDESTRUCT only in same transaction
More info here: https://eips.ethereum.org/EIPS/eip-6780

### PR DESCRIPTION
**Pull request by lorenzofero**
Closes #524. Reading the [EIP](https://eips.ethereum.org/EIPS/eip-6780), when the contract is not created in the same tx, `SELFDESTRUCT` simply becomes a transfer, while keeping the gas cost of the opcode. As such, I tried to modify only the `JournaledState` to reflect that.

### CODE REVIEW

**Comment by rakita** [comment]
Didn't forget about this, checking things related to Verkle tree.

File: crates/revm/src/journaled_state.rs [hunk 1]
```diff
@@ -468,39 +459,55 @@ impl JournaledState {
...
+            acc.info.balance = U256::ZERO;
+            JournalEntry::AccountDestroyed {
                 address,
                 target,
                 was_destroyed: previously_destroyed,
                 had_balance: balance,
-            });
+            }
+        } else {
+            Self::touch_account(self.journal.last_mut().unwrap(), &address, acc);
```
Comment (rakita) [inline crates/revm/src/journaled_state.rs:502]: This line is not needed, `address` is considered `touched` as this is the account that we currently executing.

File: crates/revm/src/journaled_state.rs [hunk 2]
```diff
@@ -468,39 +459,55 @@ impl JournaledState {
...
+            JournalEntry::BalanceTransfer {
+                from: address,
+                to: target,
+                balance,
+            }
```
Comment (rakita) [inline crates/revm/src/journaled_state.rs:507]: We are not decrementing the balance from `address`

File: crates/revm/src/journaled_state.rs [hunk 3]
```diff
@@ -468,39 +459,55 @@ impl JournaledState {
...
-
-        // NOTE: In case that target and destroyed addresses are same, balance will be lost.
-        // ref: https://github.com/ethereum/go-ethereum/blob/141cd425310b503c5678e674a8c3872cf46b7086/core/vm/instructions.go#L832-L833
-        // https://github.com/ethereum/go-ethereum/blob/141cd425310b503c5678e674a8c3872cf46b7086/core/state/statedb.go#L449
-        if address != target {
-            let target_account = self.state.get_mut(&target).unwrap();
-            // touch target account
+
+        let acc = if address != target {
+            let [acc, target_account] = self.state.get_many_mut([&address, &target]).unwrap();
```
[the remark this answers is not in the document]
Comment (rakita) [reply crates/revm/src/journaled_state.rs:479]: Oh forget this, `Hashbrown` has it

File: crates/revm/src/journaled_state.rs [hunk 4]
```diff
@@ -468,39 +459,55 @@ impl JournaledState {
...
-            .push(JournalEntry::AccountDestroyed {
+        let balance = acc.info.balance;
+        let previously_destroyed = acc.is_selfdestructed();
+        let is_cancun_enabled = SpecId::enabled(self.spec, CANCUN);
+
+        // EIP-6780 (Cancun hard-fork): selfdestruct only if contract is created in the same tx
+        let journal_entry = if acc.is_created() || !is_cancun_enabled {
+            acc.mark_selfdestruct();
+            acc.info.balance = U256::ZERO;
+            JournalEntry::AccountDestroyed {
```
Comment (rakita) [inline crates/revm/src/journaled_state.rs:495]: For JournalEntry, there is some work needed to support cancun balance change

**Review by rakita — APPROVED** [review]
good direction, left comments

**Review by rakita — CHANGES_REQUESTED** [review]
Meant to request changes

File: crates/revm/src/journaled_state.rs (hunk 2 above)
Comment (lorenzofero) [reply crates/revm/src/journaled_state.rs:507]: Oof what a dumb error, thanks for noticing

File: crates/revm/src/journaled_state.rs (hunk 4 above)
Comment (lorenzofero) [reply crates/revm/src/journaled_state.rs:495]: I'm sorry for the questions but I don't understand what you mean: do we need another enum member for `JournaledState` to track the "cancun balance change"? Why is it needed?

Comment (rakita) [reply crates/revm/src/journaled_state.rs:495]: Nothing to be done there, I thought that we needed to add the modifications for `address != target` as it has a different behaviour but that path is converted with the `JournalEntry::BalanceTransfer`.

Comment (lorenzofero) [reply crates/revm/src/journaled_state.rs:495]: Ok! Thank you for the explanation!

**Comment by lorenzofero** [comment]
I should have resolved all your comments, thanks for the support. I'm happy to see that tests are passing, even if in local they were failing for some reason..
Note that now if cancun is enabled and `address === target`, the `BalanceTransfer` entry in the JournledState is not pushed in since it is useless. As the EIP says, in such case there is no net change of balances. 
Let me know what you think!

File: crates/revm/src/journaled_state.rs [hunk 5]
```diff
@@ -468,39 +460,61 @@ impl JournaledState {
...
-
-        // NOTE: In case that target and destroyed addresses are same, balance will be lost.
-        // ref: https://github.com/ethereum/go-ethereum/blob/141cd425310b503c5678e674a8c3872cf46b7086/core/vm/instructions.go#L832-L833
-        // https://github.com/ethereum/go-ethereum/blob/141cd425310b503c5678e674a8c3872cf46b7086/core/state/statedb.go#L449
-        if address != target {
-            let target_account = self.state.get_mut(&target).unwrap();
-            // touch target account
+
+        let acc = if address != target {
+            let [acc, target_account] = self.state.get_many_mut([&address, &target]).unwrap();
```
Comment (rakita) [inline crates/revm/src/journaled_state.rs:482]:
```suggestion
            // Both accounts are loaded before this point, `address` as we execute its contract.
            // and `target` at the beginning of the function.
            let [acc, target_account] = self.state.get_many_mut([&address, &target]).unwrap();
```

File: crates/revm/src/journaled_state.rs [hunk 6]
```diff
@@ -468,39 +460,61 @@ impl JournaledState {
...
+            })
+        } else if address != target {
+            acc.info.balance = U256::ZERO;
+            Some(JournalEntry::BalanceTransfer {
+                from: address,
+                to: target,
+                balance,
+            })
+        } else {
+            None
```
Comment (rakita) [inline crates/revm/src/journaled_state.rs:512]:
```suggestion
            // State is not changed:
            // * if we are after Cancun upgrade
            // * Selfdestruct account that is created in the same transaction
            // * Specify the target to be the same as selfdestructed account. The balance stays unchanged.
            None
```

**Review by rakita — APPROVED** [review]
lgtm!
Left nit comments to add doc

**Comment by rakita** [comment]
> I should have resolved all your comments, thanks for the support. I'm happy to see that tests are passing, even if in local they were failing for some reason.. Note that now if cancun is enabled and `address === target`, the `BalanceTransfer` entry in the JournledState is not pushed in since it is useless. As the EIP says, in such case there is no net change of balances. Let me know what you think!

Tests were failing for some other reason, found the problem and merged in on `main`, so we are good

**Review by Rjected — APPROVED** [review]
other than the current comments, this looks good to me, will test this with hive once merged!

**Pull request merged**